Implement initial functionality

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR implements the initial features of this project:
* Adding permission sets to a user-specified set of Single Sign-On (SSO) groups in a user-specified set of AWS accounts
* Removing permission sets from a user-specified set of Single Sign-On (SSO) groups in a user-specified set of AWS accounts
* Adding permissions sets to a user-specified set of SSO users in a user-specified set of AWS accounts
* Removing permissions sets from a user-specified set of SSO users in a user-specified set of AWS accounts

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

When new AWS accounts are created for COOL assessment environments, one of the first steps we do is modify the SSO permissions to our preferences.  This project will be used to automate that step, which was previously a manual process.

This PR is related to https://github.com/cisagov/cool-accounts/pull/95.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I successfully applied this Terraform in the Staging and Production environments and verified that the SSO permissions were correctly modified.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

dav3r committed:
Remove examples directory

dav3r committed:
Set up S3 backend

dav3r committed:
Remove unused file inherited from skeleton

dav3r committed:
Add a remote state for the Master account

dav3r committed:
Define initial variables

dav3r committed:
Set up necessary providers

dav3r committed:
Replace skeleton outputs with a useful output

dav3r committed:
Define useful locals

dav3r committed:
Do the SSO magic

Namely:
* For each account (in local.accounts_to_configure) give each group (in 
var.groups_to_add_access_to) access to the specified permission sets
* Remove specified permission sets from users (specified in 
local.users_to_remove_access_from) that were not created by Terraform 
(such as those automatically created by AWS Control Tower)

dav3r committed:
Add a variable with the name of the SSO administration profile to use

Also, use this variable when setting up the default provider.

dav3r committed:
Make the command to delete permission sets from a user more robust

It now loops until the request status is either SUCCEEDED or FAILED.  
Also, it exits with an appropriate return code so that Terraform knows 
whether or not the local-exec command succeeded.

dav3r committed:
Add a description to the sso_group_assignments output

dav3r committed:
Add brackets around bare URLs

This avoids triggering the MD034 ("Bare URL used") linter rule when 
generating README.md.

dav3r committed:
Update README from skeleton and add useful info

dav3r committed:
Mention that jq must be installed to use this project

jsf9k commented on outputs.tf:
```diff
@@ -1,24 +1,4 @@
-  description = "The ID of the subnet where the EC2 instance is deployed."
+output "sso_group_assignments" {
+  value       = aws_ssoadmin_account_assignment.group
+  description = "The permission set assignments of each SSO groups in each AWS account."
```
```suggestion
  description = "The permission set assignments of each SSO group in each AWS account."
```

jsf9k commented on variables.tf:
```diff
@@ -3,31 +3,42 @@
+  default     = []
+}
+
+variable "users_to_remove_access_from" {
+  type        = list(object({ username = string, permission_sets = list(string) }))
+  description = "A list of objects specifying Single Sign-On (SSO) users to remove permissions from.  Each object contains the SSO username and the list of permission sets to remove access from.  Example: [{ username = \"john.doe@example.com\", permission_sets = [\"AWSAdministratorAccess\"] }]"
+  default     = []
+}
```
The code you have works for our purposes, but I could imagine a case where we might want to add access fro a particular user or remove access for a particular group.  Would it be difficult to have variables `groups_to_grant_access` and `users_to_grant_access` and
- Grant access for any groups or users that are listed
- Remove access for any groups or users that are not listed

I think this would
- Make the code more generally useful
- Give us two variables with similar intent, as opposed to two variables that are very similar but have opposite intent.  The latter is slightly confusing.

dav3r replied:
In order to correctly grant/revoke access for each user and group, we need to be able to determine which users/groups currently have access to each permission set, but Terraform doesn't give us an easy way to fetch that info.  Terraform doesn't even have a way to grab all of the existing SSO user and group names.  It would have to be accomplished via AWS CLI calls and even then it will probably get messy (messier than this already is).

As a step towards what you're asking for, how would you feel about the addition of two more variables to cover all four cases?
* `groups_to_add_access_to`
* `groups_to_remove_access_from` (new)
* `users_to_add_access_to` (new)
* `users_to_remove_access_from`

jsf9k replied:
I would be fine with that, as long as it doesn't cause you a lot of extra work.

dav3r replied:
Doing it this way is much less work than what you proposed- I don't think it will be very different from what I already have.

dav3r replied:
I added the new features and re-requested your review.  Let me know what you think.

jsf9k replied:
Thanks for doing that.  Looks great!

jsf9k reviewed (commented):
I made a suggestion that I think is worthy of your consideration.  I'm happy to approve if you respond explaining why the suggestion is a bad idea or is difficult to implement.

dav3r commented on README.md:
```diff
@@ -51,41 +73,37 @@ No modules.
-| id | The EC2 instance ID. |
-| private\_ip | The private IP of the EC2 instance. |
-| subnet\_id | The ID of the subnet where the EC2 instance is deployed. |
+| sso\_group\_assignments | The permission set assignments of each SSO groups in each AWS account. |
```
```suggestion
| sso\_group\_assignments | The permission set assignments of each SSO group in each AWS account. |
```

dav3r committed:
Correct grammatical error

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r committed:
Remove an extra blank line

dav3r committed:
Add new variables groups_to_remove_access_from and users_to_add_access_to

dav3r committed:
Update locals to use groups_to_remove_access_from and users_to_add_access_to

dav3r committed:
Add SSO functionality for users_to_add_access_to and groups_to_remove_access_from

dav3r committed:
Split SSO Terraform into multiple files to improve readability

dav3r committed:
Merge branch 'first-commits' of github.com:cisagov/cool-configure-aws-account into first-commits

dav3r committed:
Update README to reflect recent additions

dav3r committed:
Add an additional output and update README

jsf9k reviewed (approved):
Thanks for adding the extra logic.

mcdonnnj commented on providers.tf:
```diff
@@ -1,13 +1,23 @@
-#   region = var.aws_region
-# }
+# This is the "default" provider that is used to manage resources
+# inside the AWS account used for Single Sign On (SSO).
```
```suggestion
# inside the AWS account used for Single Sign-On (SSO).
```

mcdonnnj commented on sso.tf:
```diff
@@ -0,0 +1,10 @@
+data "aws_ssoadmin_permission_set" "all" {
+  for_each = local.all_permission_sets
+
+  instance_arn = tolist(data.aws_ssoadmin_instances.current.arns)[0]
```
```suggestion
  # The arns value is currently a set containing a single item
  instance_arn = tolist(data.aws_ssoadmin_instances.current.arns)[0]
```

mcdonnnj commented on sso_group.tf:
```diff
@@ -0,0 +1,44 @@
+data "aws_identitystore_group" "all" {
+  for_each = local.all_groups
+
+  identity_store_id = tolist(data.aws_ssoadmin_instances.current.identity_store_ids)[0]
```
```suggestion
  # The identity_store_ids value is currently a set containing a single item
  identity_store_id = tolist(data.aws_ssoadmin_instances.current.identity_store_ids)[0]
```

mcdonnnj commented on sso_user.tf:
```diff
@@ -0,0 +1,44 @@
+data "aws_identitystore_user" "all" {
+  for_each = local.all_usernames
+
+  identity_store_id = tolist(data.aws_ssoadmin_instances.current.identity_store_ids)[0]
```
```suggestion
  # The identity_store_ids value is currently a set containing a single item
  identity_store_id = tolist(data.aws_ssoadmin_instances.current.identity_store_ids)[0]
```

mcdonnnj commented on README.md:
```diff
@@ -2,33 +2,52 @@
+  resources, similar to
+  [this policy](https://github.com/cisagov/cool-accounts/blob/develop/master/administersso_policy.tf).
+- [Terraform](https://www.terraform.io/) installed on your system.
+- [jq](https://stedolan.github.io/jq/) installed on your system.
```
Due to the `local-exec` calls in this configuration I think we should explicitly call out the AWS CLI as a requirement.

```suggestion
- The [AWS CLI](https://aws.amazon.com/cli/) installed on your system.
- [jq](https://stedolan.github.io/jq/) installed on your system.
```

mcdonnnj reviewed (approved):
Really strong work here 💪💪💪 I just had some documentation suggestions for your consideration.

dav3r committed:
Tweak grammar in a comment

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add useful comments about Terraform SSO data

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add useful comment about Terraform SSO data

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Document AWS CLI as a prerequisite

If the AWS CLI is not installed, the local-exec calls that utilize it will fail.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Merge branch 'develop' into first-commits
